platform/cms/accelerated_publishing/rfc/broken-link-checking.md

```diff
@@ -0,0 +1,126 @@
+
+### Problems to solve
+
+In summary, the problems that the current system has are:
+
+- Content Build will no longer have complete knowledge of all pages, and so any pages that have migrated to Next Build will be flagged as broken links by Content Build
+- We will still have to test for broken links for the site as a whole.
+- Running full broken links checks at build time is too slow for an accelerated publishing scenario.
+
+## Design
+
+_Explain the proposed changes in enough detail so that a team member with working knowledge
+of the project could implement the change themselves. Bonus points if this is easily mappable
+to user stories or specs._
+
+### Link Checking Features
+
+Our system must do the following
+
+- It must operate on the entirety of va.gov content, regardless of what system the page was built by.
+- It must not create an impediment to the build process.
+- It must report the broken links into a central location that can respond to the issues.
+
+### Crawler Proposal
+
+A straight-forward way would be to create a site crawler that checks va.gov pages, follows any links found, and reports any that return a non-200/300 response. This would operate as a CI process on VA.gov's published content.
+
+All content published by Content Build and Next Build is listed in sitemap.xml files. These URLs can be read by a site crawler and the contents of the URL scanned.
+
+Advantages:
+
+- This would be relatively simple to create.
+- May be able to be largely addressed by of-the-shelf software. or entirely. [The Node.js ecosystem has several broken link checking packages](https://www.npmjs.com/search?q=broken%20link), for example.
+- Both internal and external links would be collected together into one report.
+
+Challenges:
+
+- Content would no longer be checked as it is built, which is an expectation VA.gov has due to the current architecture.
+- Determining how frequently to scan will require some experimentation.
+- This sort of checker would create additional traffic to the site. This should not present a problem to our current S3-as-web server architecture, but we should validate that the reverse proxy will not have trouble with the additional traffic, and we should ensure this traffic is ignored by analytics systems.
+
+### Build time Proposal
+
+We could check links on pages as they are built. However, these checks would need to check against the live production site, which requires making a request to a URL and verifying the response. This is _significantly_ slower than the current Content Build system.
+
+It is only recommended that we do this when we move to an incremental publishing scenario, and also that the link checks happen independently of deployment. In other words, link checks would be triggered by publishing of pages, but they would happen after deployment, rather than before.
```

Comment:
This bit makes it sound like publishing a page is the only source of the broken link.  As in, if I publish a new version of my page, I need all the links on that page checked.  In reality though, I also need all the links TO that page that exist on the site to be checked, because my content may not have added new links, but it might have changed my page location, or I unpublished a page.   So I do not think a build time solution could cover enough all on its own.

Reply:
Yeah, the broken in-links problem from archiving a page can't be handled by a check on the changed pages alone. It requires full scans. I'll update this to be clearer and to call out that specific problem.